cep-incremental-repodata.md

```diff
@@ -0,0 +1,289 @@
+Patch 2.97s
+Serialize 1.36s
+```
+
+The times are from PyPy running on an older Intel(R) Core(TM) i7-5500U laptop.
+
+## Summary
+
+The proposed system saves bandwith when a locally cached `repodata.json` is
+known by the patch server, with a very concise client implementation. After the
+first update, subsequent updates fetch only the newest information using a
+single round trip to the server.
+
+## Alternatives
+
+JSON Merge Patch is simpler but does not allow the `null` values occasionally
+used in `repodata.json`.
+
+Textual diff + patch would work, but `conda` needs the data and not the
+formatting.
+
+`zchunk` is a compression format used in Fedora, implemented in C. It splits
+files into independently compressed chunks, transferring changed chunks on
+update. It is generic on bytes. The server does not have to keep a history of
+older versions. The web server should support multipart Range requests, not true
+of s3, probably OK for CDN.
+
+## JSON Lines With Leading and Trailing Checksums
+
+```
+0000000000000000000000000000000000000000000000000000000000000000
+{"to": "af99a2269795b91aee909eebc6b71a127f8001475c67c2345c96253b97378b21", "from": "af99a2269795b91aee909eebc6b71a127f8001475c67c2345c96253b97378b21", "patch": []}
+...
+{"url": "", "latest": "af99a2269795b91aee909eebc6b71a127f8001475c67c2345c96253b97378b21", "headers": {"date": "Thu, 14 Apr 2022 17:24:36 GMT", "last-modified": "Tue, 28 May 2019 02:01:41 GMT"}}
+8b9c825f33cc68354f131dd810a068256e34153b7335619eeb187b51a54c7118
+```
+
+The `.jlap` format allows clients to fetch the newest patches with a single HTTP
+Range request. It consists of a leading checksum, any number of lines of the
+elements of the `"patches"` array and a `"metadata"` line in the [JSON
+Lines](https://jsonlines.readthedocs.io/en/latest/) format, and a trailing
+checksum.
+
+The checksums are constructed in such a way that the trailing checksum can be
+re-verified without re-reading (or retaining) the beginning of the file, if the
+client remembers an intermediate checksum.
+
+When `repodata.json` changes, the server wil truncate the `"metadata"` line,
+appending new patches, a new metadata line and a new trailing checksum.
+
+When the client wants new data, it issues a single HTTP Range request from the
+bytes offset of the beginning of the penultimate `"metadata"` line, to the end
+of the file (`Range: bytes=<offset>-`), and re-verifies the trailing checksum.
+If the trailing checksum does not match the computed checksum then it must
+re-fetch the entire file; otherwise, it may apply the new patches.
+
+If the `.jlap` file represents part of a stream (earlier lines have been
+discarded) then the leading checksum is an intermediate checksum from that
+stream. Otherwise the leading checksum is all `0`'s.
+
+## Reference
+
+* JSON Patch https://datatracker.ietf.org/doc/html/rfc6902
+* Server and client implementation in Python https://github.com/dholth/repodata-fly
```

Comment:
The methods are all slightly beyond me, but the idea sounds really great. Being able to download only part of the repodata.json would be a great way to make updates more efficient.